### ISSUE TITLE
Typescript?

### ISSUE THREAD

**Issue by Lucretiel** [TI4-Online/TI4-TTPG#1]
TTPG ships typescript definitions for their whole API; any chance we could use it for this project? I've found it to be a tremendous aid in ensuring correctness throughout a project and reducing runtime issues

**Comment by Joseph-Alley** [comment]
I'll second this motion. I find typescript incredibly useful for catching small, simple bugs, and maintaining code quality.

**Comment by RobMayer** [comment]
The problem with using Typescript, or even more recent versions of ecma (2020+) is that it requires transpilation using babel. 
 We've laid the ground work to be able to do that, but we've opted to bypass transpilation for now as it makes hooking into TTPGs debug loop much more difficult. You end up getting debug and error reporting on your transpiled code and so on.

Personally, I am not a fan of typescript, but since I am not going to be leading the scripting aspects of this project, my opinion on it matters less. I defer to @darrellanderson on this one.

**Comment by darrellanderson** [comment]
What would be involved in the transpiling process?  We'd write foo.ts and "yarn dev" + "yarn dev watch" would transpile to foo.js?

I'm not opposed to it, but for the time being would prefer to stick with what TTPG handles natively.  We can lean on "type-check" and other runtime solutions, as well as good coverage with unittests + regression testing when things slip through.

**Comment by RobMayer** [comment]
> What would be involved in the transpiling process? We'd write foo.ts and "yarn dev" + "yarn dev watch" would transpile to foo.js?

Pretty much that, yeah. I think I'd need to change a few things in the .babelrc config file, add a module or two, and enable the "transpile" flag in the config/project.json, but that's all tooling related and would be on me to sort out if/when we went that direction.

**Comment by Lucretiel** [comment]
I agree that a seamless in-editor debugging experience is a requirement. I think that TTPG uses V8, which means that it should be able to support source maps to enable this debugging. I can only assume that there's some solution for this, given that TTPG publish first-party typescript definitions for their API at https://github.com/plasticity-studios/tabletop-playground-api. I'll play around with the build pipeline and see if I can get it working.

**Comment by Lucretiel** [comment]
Now that the CI and prettier PRs are finished, I'm targeting next week for an initial version of this. Some notes about earlier concerns:

- We should be able to entirely use `tsc` instead of `babel`, especially because we aren't currently doing any transpiling to begin with. `tsc` supports basic source transformations in addition to the typescript -> javascript conversions (for instance, converting `let / const` to `var`, arrow functions to `function`, etc.)
- It appears that TTPG doesn't currently have strong support for source-level debugging to begin with, unless I'm missing something, so there isn't much concern about that. In any case, because they use V8, we'll be able to use source maps to ensure a good source-level debugging experience once TTPG has better support.
- typescript (through the `any` type) support a smooth transition from javascript. This means that I'll be able to introduce the relevant tooling (configure typescript builds, watch mode, CI, etc), and convert one of the js files to ts as a proof-of-concept, but leave the other ones untouched for now. Once it's all working and merged I'll go through in follow-up PRs converting other JS files.
